### PR TITLE
GEODE-4647: add a stats eventSecondaryQueueSizeId to track events in …

### PR DESCRIPTION
**Pull request by gesterzhou**
…secondary

            gateway sender queue

@boglesby @jhuynh1 

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/cache/BucketAdvisor.java [hunk 1]
```diff
@@ -282,8 +282,11 @@ public boolean deposePrimary() {
           if (this.getBucket() instanceof BucketRegionQueue) {
             BucketRegionQueue brq = (BucketRegionQueue) this.getBucket();
             brq.decQueueSize(brq.size());
+            brq.incSecondaryQueueSize(brq.size());
```
Comment (jhuynh1) [inline geode-core/src/main/java/org/apache/geode/internal/cache/BucketAdvisor.java]: Should this be in the else block?

File: geode-core/src/main/java/org/apache/geode/internal/cache/AbstractRegionMap.java [hunk 2]
```diff
@@ -1042,6 +1042,10 @@ public boolean initialImagePut(final Object key, final long lastModified, Object
           } finally {
             if (done && result) {
               initialImagePutEntry(newRe);
+              if (owner instanceof BucketRegionQueue) {
```
Comment (jhuynh1) [inline geode-core/src/main/java/org/apache/geode/internal/cache/AbstractRegionMap.java]: Why is this needed here?  Are we missing events currently?

File: geode-core/src/main/java/org/apache/geode/internal/cache/BucketAdvisor.java [hunk 3]
```diff
@@ -1192,6 +1195,7 @@ protected boolean acquiredPrimaryLock() {
           if (br instanceof BucketRegionQueue) { // Shouldn't it be AbstractBucketRegionQueue
             BucketRegionQueue brq = (BucketRegionQueue) br;
             brq.incQueueSize(brq.size());
+            brq.decSecondaryQueueSize(brq.size());
```
Comment (jhuynh1) [inline geode-core/src/main/java/org/apache/geode/internal/cache/BucketAdvisor.java]: Why do we always decrement the secondary queue size?  It looks like we don't add it unless it's a secondary from below.

File: geode-wan/src/test/java/org/apache/geode/internal/cache/wan/WANTestBase.java [hunk 4]
```diff
@@ -3131,10 +3210,10 @@ public static Integer getQueueContentSize(final String senderId, boolean include
         return ((ParallelGatewaySenderQueue) regionQueue).localSize(includeSecondary);
       } else {
         if (includeSecondary) {
-          fail("Not Implemented yet");
+          fail("Not implemented yet");
```
Comment (jhuynh1) [inline geode-wan/src/test/java/org/apache/geode/internal/cache/wan/WANTestBase.java]: This test should be failing if this line is present... the below lines would never be called

File: geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANConflationDUnitTest.java [hunk 5]
```diff
@@ -108,24 +113,35 @@ public void testParallelPropagationBatchConflation() throws Exception {
       vm4.invoke(() -> putGivenKeyValue(getTestMethodName(), keyValues));
     }
 
+    // sender did not turn on conflation, so queue size will be 100 (otherwise it will be 20)
```
Comment (jhuynh1) [inline geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANConflationDUnitTest.java]: This test is named conflation, is there a reason why it is now getting 100 instead of 20?

File: geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANStatsDUnitTest.java [hunk 6]
```diff
@@ -51,6 +52,172 @@ protected final void postSetUpWANTestBase() throws Exception {
...
+    createCacheInVMs(nyPort, vm2);
+    createReceiverInVMs(vm2);
+
+    createSendersWithConflation(lnPort);
+
+    createSenderPRs(1);
+
+    startPausedSenders();
+
+    createReceiverPR(vm2, 1);
```
Comment (jhuynh1) [inline geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANStatsDUnitTest.java]: Should probably create receivers first if possible.  I understand that we are pausing the senders so we are "safe" here but I think most tests should create receivers first

File: geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANStatsDUnitTest.java [hunk 7]
```diff
@@ -51,6 +52,172 @@ protected final void postSetUpWANTestBase() throws Exception {
...
+    ArrayList<Integer> v5List =
+        (ArrayList<Integer>) vm5.invoke(() -> WANTestBase.getSenderStats("ln", NUM_PUTS));
+    ArrayList<Integer> v6List =
+        (ArrayList<Integer>) vm6.invoke(() -> WANTestBase.getSenderStats("ln", NUM_PUTS));
+    ArrayList<Integer> v7List =
+        (ArrayList<Integer>) vm7.invoke(() -> WANTestBase.getSenderStats("ln", NUM_PUTS));
+
+    assertEquals(NUM_PUTS, v4List.get(0) + v5List.get(0) + v6List.get(0) + v7List.get(0)); // queue
+                                                                                           // size
+    assertEquals(NUM_PUTS * 2, v4List.get(1) + v5List.get(1) + v6List.get(1) + v7List.get(1)); // eventsReceived
```
Comment (jhuynh1) [inline geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANStatsDUnitTest.java]: How are we receiving events if the sender is never resumed?

File: geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANStatsDUnitTest.java [hunk 8]
```diff
@@ -51,6 +52,172 @@ protected final void postSetUpWANTestBase() throws Exception {
...
+    assertEquals(NUM_PUTS * 2, v4List.get(2) + v5List.get(2) + v6List.get(2) + v7List.get(2)); // events
+                                                                                               // queued
+    assertEquals(0, v4List.get(3) + v5List.get(3) + v6List.get(3) + v7List.get(3)); // events
+                                                                                    // distributed
+    assertEquals(NUM_PUTS, v4List.get(10) + v5List.get(10) + v6List.get(10) + v7List.get(10)); // secondary
+                                                                                               // queue
+                                                                                               // size
+
+    // stop vm7 to trigger rebalance and move some primary buckets
```
Comment (jhuynh1) [inline geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANStatsDUnitTest.java]: Perhaps move this testing to a different test.  So possibly one test to check queue size, another to do rebalance and check queue size, etc.  Makes debugging future failures a little easier

File: geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANStatsDUnitTest.java [hunk 9]
```diff
@@ -51,6 +52,172 @@ protected final void postSetUpWANTestBase() throws Exception {
...
+    v7List = (ArrayList<Integer>) vm7.invoke(() -> WANTestBase.getSenderStats("ln", 0));
+
+    assertEquals(NUM_PUTS, v4List.get(3) + v5List.get(3) + v6List.get(3) + v7List.get(3)); // events
+                                                                                           // distributed
+    assertEquals(0, v4List.get(10) + v5List.get(10) + v6List.get(10) + v7List.get(10)); // secondary
+                                                                                        // queue
+                                                                                        // size
+  }
+
+  // TODO: add a test without redudency for primary switch
```
Comment (jhuynh1) [inline geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANStatsDUnitTest.java]: remove todo?

File: geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANStatsDUnitTest.java [hunk 10]
```diff
@@ -51,6 +52,172 @@ protected final void postSetUpWANTestBase() throws Exception {
...
+    // sizes:"+v4List.get(10)+":"+v5List.get(10)+":"+v6List.get(10)+":"+v7List.get(10));
+    // System.out.println("Now move a primary bucket");
+    // final DistributedMember vm6member = vm6.invoke(() -> WANTestBase.getMember());
+    // vm7.invoke(() -> WANTestBase.moveAllPrimaryBuckets("ln", vm6member, testName));
+    //
+    // v4List = (ArrayList<Integer>) vm4.invoke(() -> WANTestBase.getSenderStats("ln", NUM_PUTS));
+    // v5List = (ArrayList<Integer>) vm5.invoke(() -> WANTestBase.getSenderStats("ln", NUM_PUTS));
+    // v6List = (ArrayList<Integer>) vm6.invoke(() -> WANTestBase.getSenderStats("ln", NUM_PUTS));
+    // v7List = (ArrayList<Integer>) vm7.invoke(() -> WANTestBase.getSenderStats("ln", NUM_PUTS));
+    // assertEquals(NUM_PUTS, v4List.get(10) + v5List.get(10) + v6List.get(10) + v7List.get(10)); //
```
Comment (jhuynh1) [inline geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANStatsDUnitTest.java]: remove commented code

File: geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/AsyncEventQueueStats.java [hunk 11]
```diff
@@ -42,6 +42,8 @@
             f.createLongCounter(EVENT_QUEUE_TIME, "Total time spent queueing events.",
                 "nanoseconds"),
             f.createIntGauge(EVENT_QUEUE_SIZE, "Size of the event queue.", "operations", false),
+            f.createIntGauge(EVENT_SECONDARY_QUEUE_SIZE, "Size of the event secondary queue.",
```
Comment (jhuynh1) [inline geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/AsyncEventQueueStats.java]: Maybe "Size of the secondary event queue" or "Number of secondary events in queue"

File: geode-core/src/main/java/org/apache/geode/internal/cache/BucketAdvisor.java (hunk 1 above)
Comment (gesterzhou) [reply geode-core/src/main/java/org/apache/geode/internal/cache/BucketAdvisor.java]: No, it must be this way. When a member turned from primary to secondary, it increased secondary queue size, and decreased primary queue size, together. 

I have a test case with redundency=1, however, I'm thinking to add a dunit test with redundency=0, which requires some tricky code.

File: geode-core/src/main/java/org/apache/geode/internal/cache/AbstractRegionMap.java (hunk 2 above)
Comment (gesterzhou) [reply geode-core/src/main/java/org/apache/geode/internal/cache/AbstractRegionMap.java]: Yes. This is GII logic. I talked to you last Friday.

File: geode-core/src/main/java/org/apache/geode/internal/cache/BucketAdvisor.java (hunk 3 above)
Comment (gesterzhou) [reply geode-core/src/main/java/org/apache/geode/internal/cache/BucketAdvisor.java]: If reduency equals to 1, I have a test case to prove its necessary. I will add one test case with no secondary. Then I want to see if it has any problems.

File: geode-wan/src/test/java/org/apache/geode/internal/cache/wan/WANTestBase.java (hunk 4 above)
Comment (gesterzhou) [reply geode-wan/src/test/java/org/apache/geode/internal/cache/wan/WANTestBase.java]: The whole "else" will never be executed. I have confirmed with Barry. 

I will change it to be:
} else {
  fail("Not Implemented yet");
}

File: geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANConflationDUnitTest.java (hunk 5 above)
Comment (gesterzhou) [reply geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANConflationDUnitTest.java]: The test is for batch conflation, and purposely not to turn on the conflation during enqueue. So the queue size in this test will be 100.  

We have other cases to turn on conflation during enqueue, where the queue size will be 20.

File: geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANStatsDUnitTest.java (hunk 7 above)
Comment (gesterzhou) [reply geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANStatsDUnitTest.java]: This has nothing to do with my code changes. However, the eventsReceivedId is misleading somewhat. 

From the context, the sender.distribute(), before enqueue,  it will call incEventsReceived(). 
It sounds like to indicate an event received by the sender (not even enqueue yet). 

If you think it's a bug, we can log another ticket for it.

File: geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANStatsDUnitTest.java (hunk 9 above)
Comment (gesterzhou) [reply geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANStatsDUnitTest.java]: Something I intend to add.

File: geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/AsyncEventQueueStats.java (hunk 11 above)
Comment (gesterzhou) [reply geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/AsyncEventQueueStats.java]: I just want to follow the same term with "Size of the event queue"

File: geode-core/src/main/java/org/apache/geode/internal/cache/AbstractRegionMap.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (gesterzhou) [reply geode-core/src/main/java/org/apache/geode/internal/cache/AbstractRegionMap.java]: I have to handle that together, otherwise, test case to change primary bucket will fail. I have a test case for it.

File: geode-core/src/main/java/org/apache/geode/internal/cache/BucketAdvisor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (gesterzhou) [reply geode-core/src/main/java/org/apache/geode/internal/cache/BucketAdvisor.java]: I have similar concern too. I'm adding a test case for reducent=0
